lib/schema.js

```diff
@@ -129,21 +129,75 @@ const linkSchemas = (schemas, categories, domains) => {
   return errors.length ? new MetaschemaError(errors) : null;
 };
 
-const extractDefinition = (definition, path, ms) => {
+const extractDefinition = (definition, path, ms, prefix, info) => {
+  let property = path;
   const separatorIndex = path.indexOf('.');
+  if (separatorIndex !== -1) {
+    property = path.slice(0, separatorIndex);
+  }
+  if (!definition[property]) {
+    throw new SchemaValidationError(
+      'unresolvedProperty',
+      info.source,
+      info.property,
+      { property: `${prefix}${property}` }
+    );
+  }
+
   if (separatorIndex === -1) {
-    return definition[path];
+    return definition[property];
   }
-  const property = path.slice(0, separatorIndex);
+
   const categoryName = definition[property].category;
-  const category = ms.categories.get(categoryName).definition;
-  return extractDefinition(category, path.slice(separatorIndex + 1), ms);
+  const category = ms.categories.get(categoryName);
+
+  return extractDefinition(
+    category.definition,
+    path.slice(separatorIndex + 1),
+    ms,
+    `${prefix}${categoryName}.`,
+    info
+  );
+};
+
+const extractByPath = (definition, path, ms, info) => {
+  const isOuterCategory = path.startsWith('::');
+  if (!isOuterCategory) {
+    return extractDefinition(definition, path, ms, '', info);
+  }
+
+  const separatorIndex = path.indexOf('.');
+  const categoryName =
+    separatorIndex === -1 ? path.slice(2) : path.slice(2, separatorIndex);
+  const category = ms.categories.get(categoryName);
+
+  const endOfCategory = separatorIndex === -1 ? path.length : separatorIndex;
+  if (!category) {
+    throw new SchemaValidationError(
+      'unresolvedCategory',
+      info.source,
+      info.property,
+      { category: path.slice(0, endOfCategory) }
+    );
+  }
+
+  return separatorIndex === -1 ?
+    { category: categoryName, definition: category.definition } :
```

Comment:
Can't you just return `category` here?

Reply:
Because it doesn't have `category` property.